cluster-operator/src/test/java/io/strimzi/operator/cluster/operator/assembly/KafkaRebalanceAssemblyOperatorTest.java

```diff
@@ -90,65 +99,661 @@
+                            kr1);
+                }).setHandler(context.succeeding(v -> context.verify(() -> {
+                    // the resource moved from ProposalPending to Stopped
+                    KafkaRebalance kr3 = Crds.kafkaRebalanceOperation(kubernetesClient).inNamespace(CLUSTER_NAMESPACE).withName(RESOURCE_NAME).get();
+                    assertState(kr3, KafkaRebalanceAssemblyOperator.State.Stopped);
+                }))).compose(v -> {
+
+                    // apply the "refresh" annotation to the resource in the Stopped state
+                    KafkaRebalance kr4 = Crds.kafkaRebalanceOperation(kubernetesClient).inNamespace(CLUSTER_NAMESPACE).withName(RESOURCE_NAME).get();
+
+                    KafkaRebalance refreshedKr = new KafkaRebalanceBuilder(kr4)
+                            .editMetadata()
+                                .addToAnnotations(KafkaRebalanceAssemblyOperator.ANNO_STRIMZI_IO_REBALANCE, "refresh")
+                            .endMetadata()
+                            .build();
+
+                    Crds.kafkaRebalanceOperation(kubernetesClient)
+                            .inNamespace(CLUSTER_NAMESPACE)
+                            .withName(RESOURCE_NAME)
+                            .patch(refreshedKr);
+
+                    // trigger another reconcile to process the Stopped state
+                    return kcrao.reconcileRebalance(
+                            new Reconciliation("test-trigger", KafkaRebalance.RESOURCE_KIND, CLUSTER_NAMESPACE, RESOURCE_NAME),
+                            refreshedKr);
+                }).setHandler(context.succeeding(v -> context.verify(() -> {
+                    // the resource moved from Stopped to PendingProposal
+                    KafkaRebalance kr5 = Crds.kafkaRebalanceOperation(kubernetesClient).inNamespace(CLUSTER_NAMESPACE).withName(RESOURCE_NAME).get();
+                    assertState(kr5, KafkaRebalanceAssemblyOperator.State.PendingProposal);
+                }))).compose(v -> {
+                    // trigger another reconcile to process the PendingProposal state
+                    KafkaRebalance kr6 = Crds.kafkaRebalanceOperation(kubernetesClient).inNamespace(CLUSTER_NAMESPACE).withName(RESOURCE_NAME).get();
+
+                    return kcrao.reconcileRebalance(
+                            new Reconciliation("test-trigger", KafkaRebalance.RESOURCE_KIND, CLUSTER_NAMESPACE, RESOURCE_NAME),
+                            kr6);
+                }).setHandler(context.succeeding(v -> context.verify(() -> {
+                    // the resource moved from PendingProposal to ProposalReady
+                    KafkaRebalance kr7 = Crds.kafkaRebalanceOperation(kubernetesClient).inNamespace(CLUSTER_NAMESPACE).withName(RESOURCE_NAME).get();
+                    assertState(kr7, KafkaRebalanceAssemblyOperator.State.ProposalReady);
+                    context.completeNow();
+                })));
```

Comment:
I've seeing a very large chain of compose etc here, and (I've not looked thoroughly) I suspect there's a lot of duplication between the tests in the kinds of things the handler are doing. If so there's scope to factor out the handlers as methods, which will make the tests easier to read and maintain.

Reply:
+1 to this. If we can capture some singular chunks of the compose + handlers we can turn them into re-usable pieces and the readability of each tests should go up

Reply:
I have factored out a few methods for common operations to be used in the handlers.